src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -0,0 +1,325 @@
+        }
+
+        internal static TypeName? Parse(ReadOnlySpan<char> typeName, bool throwOnError, TypeNameParseOptions? options = default)
+        {
+            ReadOnlySpan<char> trimmedName = typeName.TrimStart(); // whitespaces at beginning are always OK
+            if (trimmedName.IsEmpty)
+            {
+                // whitespace input needs to report the error index as 0
+                return throwOnError ? throw ArgumentException_InvalidTypeName(errorIndex: 0) : null;
+            }
+
+            int recursiveDepth = 0;
+            TypeNameParser parser = new(trimmedName, throwOnError, options);
+            TypeName? parsedName = parser.ParseNextTypeName(allowFullyQualifiedName: true, ref recursiveDepth);
+
+            if (parsedName is not null && parser._inputString.IsEmpty) // unconsumed input == error
+            {
+                return parsedName;
+            }
+            else if (!throwOnError)
+            {
+                return null;
+            }
+
+            if (recursiveDepth >= parser._parseOptions.MaxNodes)
+            {
+                throw InvalidOperation_MaxNodesExceeded(parser._parseOptions.MaxNodes);
+            }
+
+            int errorIndex = typeName.Length - parser._inputString.Length;
+            throw ArgumentException_InvalidTypeName(errorIndex);
+        }
+
+        // this method should return null instead of throwing, so the caller can get errorIndex and include it in error msg
+        private TypeName? ParseNextTypeName(bool allowFullyQualifiedName, ref int recursiveDepth)
+        {
+            if (!TryDive(ref recursiveDepth))
+            {
+                return null;
+            }
+
+            List<int>? nestedNameLengths = null;
+            if (!TryGetTypeNameInfo(ref _inputString, ref nestedNameLengths, out int fullTypeNameLength, out int genericArgCount))
+            {
+                return null;
+            }
+
+            ReadOnlySpan<char> fullTypeName = _inputString.Slice(0, fullTypeNameLength);
+            _inputString = _inputString.Slice(fullTypeNameLength);
+
+            int genericArgIndex = 0;
+            // Don't allocate now, as it may be an open generic type like "List`1"
+            TypeName[]? genericArgs = null;
+
+            // Are there any captured generic args? We'll look for "[[" and "[".
+            // There are no spaces allowed before the first '[', but spaces are allowed
+            // after that. The check slices _inputString, so we'll capture it into
+            // a local so we can restore it later if needed.
+            ReadOnlySpan<char> capturedBeforeProcessing = _inputString;
+            if (IsBeginningOfGenericAgs(ref _inputString, out bool doubleBrackets))
+            {
+                int startingRecursionCheck = recursiveDepth;
+                int maxObservedRecursionCheck = recursiveDepth;
+
+            ParseAnotherGenericArg:
+
+                // Invalid generic argument count provided after backtick.
```

Comment:
The number after backtick is just a CLS convention. It is perfectly fine to have types that do not follow it. (Obfuscators like to create types with names like this to complicate decompilation.)

A compliant type name parser should not depend on this convention. Nothing in the runtime type loader depends on this convention today. For example, the following test passes before this change, but fails with this change:
```
using System;
using System.Reflection;
using System.Reflection.Emit;

var ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.Run);
var mb = ab.DefineDynamicModule("MyModule");
var tb = mb.DefineType("MyDynamicType`12345", TypeAttributes.Public);
tb.DefineGenericParameters(["T"]);
var t = tb.CreateType();
var ti = t.MakeGenericType(typeof(int));
Console.WriteLine(t.Assembly.GetType(ti.FullName, throwOnError: true).ToString());
```

Reply:
Great catch, I was not aware of that (and I think you have just discovered another test coverage hole).

> A compliant type name parser should not depend on this convention. 

I agree, but on the other hand I believe that we should have the possibility to reject such inputs.

cc @GrabYourPitchforks

Reply:
> I agree, but on the other hand I believe that we should have the possibility to reject such inputs.

In the binary formatter scenario, would not this input be naturally rejected by any allow list anyway?

Reply:
> would not this input be naturally rejected by any allow list anyway?

Not necessarily. I could easily imagine somebody doing the following:

```cs
if (typeName.Name == "System.Collections.Generic.List`1"
    && typeName.Assembly == "mscorlib, ..."
    && typeName.GetGenericArguments()[0].Name == "Contoso.TrustedModels.Foo")
{
    // passed all checks - allow it through!
    Type resolved = typeName.Resolve();
}
```

And then somebody submits to the server:

```txt
System.Collections.Generic.List`1[[Contoso.TrustedModels.Foo], [Attacker.Dangerous.Haha]], mscorlib, ...
```

What does the `Resolve` method above do in this case? Does it start resolving inner types before realizing that the outermost type (`List<T>`) only takes a single generic arg?

Reply:
That looks like a bug in the allow list implementation to me.

Reply:
> That looks like a bug in the allow list implementation to me.

I agree it's a bug. But is it the type of bug that somebody might reasonably make because they don't know all the nuances of how the runtime handles generics? I think the answer is yes: most developers would be surprised to learn that the things they believed to be universally true based on their experience with .NET and C# are merely unenforced conventions.

Reply:
Developers writing the hardened resolvers against this object model need to be aware of the nuances and/or write a very defensive code. I do not see a way around it.

I think that the parser should not be doing any validation of the structural integrity of the TypeName tree. For example, `byref of byref` happens to be invalid in C# and .NET today (but it may change at some point in future - there is nothing fundamental preventing `byref of byref` from existing). Should the type name parse reject it? Is it ok for people building hardened to depend on parser rejecting it? I think the answers to both of these questions should be negative.

Reply:
Then I think what we'll want to do for this library is document that it doesn't make any assumptions. Maybe something like: this library provides a faithful parsed representation of the input string, but it's not opinionated and doesn't enforce standard .NET conventions, and callers who need this must perform these checks themselves.

That would allow moving all of the policy checks (no path traversal in the assembly name, no escape characters in type names, generic arity markers must follow normal conventions, no ref to ref, etc.) in a higher-level package which depends on this low-level library. And if the SDL crew wants to mandate use of this library internally for apps & services, they'd mandate use of the opinionated higher-level package. I think there's precedent for this type of split.

(Note: the opinionated high-level library would not be owned by the Libraries team. There are other teams who would create and maintain it for internal use, perhaps consulting with us as needed.)